olive/passes/onnx/moe_experts_distributor.py

```diff
@@ -0,0 +1,259 @@
+    for expert in experts:
+        for i, j in [(0, 1), (-2, -1)]:
+            prev_node = nodes[expert[i]]
+            next_node = nodes[expert[j]]
+
+            node_name = _create_node_name(model, "AllToAll", prev_node.name, next_node.name)
+            output_name = node_name + "_output_0"
+            alltoall = onnx.helper.make_node(
+                "AllToAll", prev_node.output, [output_name], name=node_name, domain=_domain, group_size=world_size
+            )
+
+            model.add_node(alltoall)
+            OnnxModel.replace_node_input(next_node, prev_node.output[0], output_name)
+
+
+def _replace_constant_value(constant: Message, value: int):
+    for attr in constant.attribute:
+        if attr.name == "value":
+            attr.CopyFrom(
+                onnx.helper.make_attribute(
+                    attr.name, onnx.numpy_helper.from_array(numpy.array([value], numpy.int64)), attr.doc_string
+                )
+            )
+            return
+
+
+def _fix_shapes(
+    model: OnnxModel,
+    nodes: Dict[str, Message],
+    producers: Dict[str, Message],
+    consumers: Dict[str, Message],
+    experts: List[List[str]],
+    world_size: int,
+    num_experts: int,
+    rank: int,
+):
+    experts_per_rank = int(num_experts / world_size)
+
+    for expert in experts:
+        div_node = nodes[expert[5]]
+        div_node_iname = div_node.input[1]
+        div_node_oname = div_node.output[0]
+        _replace_constant_value(producers[div_node_iname], experts_per_rank)
+
+        add_node = nodes[expert[4]]
+        add_node_iname = add_node.input[1]
+        _replace_constant_value(producers[add_node_iname], experts_per_rank - 1)
+
+        mul_nodes = [n for n in consumers[div_node_oname] if n.op_type == "Mul"]
+        index = 1
+        for mul_node in mul_nodes:
+            if div_node_oname in mul_node.input:
+                _replace_constant_value(producers[mul_node.input[1]], index)
+                index += 1
+
+        reshape_node = nodes[expert[1]]
+        for parent in model.get_parents(reshape_node, producers):
+            if parent.op_type == "Concat":
+                concat_iname_0 = parent.input[0]
+                _replace_constant_value(producers[concat_iname_0], int(num_experts / experts_per_rank))
+
+                concat_iname_1 = parent.input[1]
+                _replace_constant_value(producers[concat_iname_1], experts_per_rank)
+
+                break
+
+
+def run(world_size: int, input_filepath: str, output_dirpath: str, debug: bool = False):
+    basename = Path(input_filepath).stem
+    output_dirpath = Path(output_dirpath)
+
+    model = OnnxModel(onnx.load_model(input_filepath))
+    producers = model.output_name_to_node()
+    consumers = model.input_name_to_nodes()
+
+    if debug:
+        OnnxModel.graph_topological_sort(model.model.graph)
+        _dump_graph(model.model, str(output_dirpath / f"{basename}.graph"))
```

Comment:
is there a reason you implemented `_dump_graph` instead of saving the model using `onnx.save_model`?

Reply:
or directly call `model.save_model_to_file` which also does the topological sort inside it? https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/onnx_model.py#L1047

Reply:
`_dump_graph` is in text format (basically, json) not the usual binary onnx format.

Reply:
Oh I see. Does it make it more readable?

Reply:
It's for debugging purposes and that's why wrapped with `debug` flag. It's simpler to compare to json objects than visually comparing two graphs in tool like netron. It has been handy.

Reply:
That sounds good to me. Was wondering what the differences were. Thanks!